Initial pass to improve token documentation

maririos opened the pull request:
The main goal behind the PR is to use the classes as bases for documentation. I am mainly organizing information.
I don't intend to change code here. Comments related to code will be in original PR https://github.com/Azure/azure-sdk-tools/pull/7980

maririos commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -6,44 +6,138 @@
+        /// <description>GroupId: `doc` to group consecutive comment tokens as documentation.</description>
+        /// </item>
+        /// <item>
+        /// <description>NavigateToId: Id for navigating to where the node where the token is defined. Should match the definitionId of the referenced node.</description>
```
@chidozieononiwu should this be `should match the Idof the referenced node` instead of `definitionId` ?

tjprescott commented on tools/apiview/parsers/CONTRIBUTING.md:
```diff
@@ -45,66 +57,32 @@ Each tree node has top tokens which should be used to capture the main tokens on
-
-Serialize the generated code file to JSON them compress the file using Gzip compression. Try to make the json as small as possible by ignoring null values and empty collections.
+Serialize the generated code file to JSON and then compress the file using Gzip compression. Try to make the json as small as possible by ignoring null values and empty collections.
+Don't worry about indentation that will be handled by the tree structure. In the case you want to have indentation between the tokens then use `TabSpace` token kind.
```
This statement is not true. Indentation is only "handled for you" when you model things with parent child relationships. Vast amounts of information are expected to be put into "TopTokens" where you will get no such automatic indentation.

chidozieononiwu replied:
But isn't that what it says in the later part of the paragraph. Perhaps I can make it clearer.

tjprescott commented on tools/apiview/parsers/CONTRIBUTING.md:
```diff
@@ -45,66 +57,32 @@ Each tree node has top tokens which should be used to capture the main tokens on
+## Serialization
 
-### StructuredToken
-- `Value` : The token value which will be dispalyed. Spacing tokens don't need to have value.
```
Unless the plan is to actually render the doc comments to markdown, removing this information is a net negative. Parser authors should not have to be looking through C# code for documentation.

maririos replied:
Understood. I am looking into it. But this way the documentation keeps updated with the code. We will keep improving this

tjprescott commented on tools/apiview/parsers/CONTRIBUTING.md:
```diff
@@ -45,66 +57,32 @@ Each tree node has top tokens which should be used to capture the main tokens on
-- PUNCTUATION : Create a token with `Kind = Content` and the `Value = the punctuation`
+
+- TEXT, KEYWORD, COMMENT : Use `text`, `keyword`, `comment` to property `RenderClasses` of `StructuredToken`.
+- NEW_LINE : Create a token with `Kind = LineBreak`.
+- WHITE_SPACE :  Create token with `Kind = NoneBreakingSpace`.
+- PUNCTUATION : Create a token with `Kind = Content` and the `Value = <the punctuation>`.
 - DOCUMENTATION : Add `GroupId = doc` in the properties of the token. This identifies a range of consecutive tokens as belonging to a group.
-- SKIP_DIFF :  Add `SkipDiff` to the Tag to indicate that node or token should not be included in diff computation
+- SKIP_DIFF :  Add `SkipDiff` to the Tag to indicate that the node or token should not be included in diff computation.
 - LINE_ID_MARKER : You can add a empty token. `Kind = Content` and `Value = ""` then give it an `Id` to make it commentable.
 - EXTERNAL_LINK : Create a single token set `Kind = Url`, `Value = link` then add the link text as a properties `LinkText`;
 - Common Tags: `Deprecated`, `Hidden`, `HideFromNav`, `SkipDiff`
 - Cross Language Id: Use `CrossLangId` as key with value in the node properties.
```
These should probably be little subsections, and show the necessary JSON payload. That's the only cross-language representation there is.

maririos replied:
We can add more information about cross language when the feature is fully implemented

tjprescott reviewed (commented):
Comments. Many imply code changes, which I do understand isn't the focus of this PR. But the code changes would result in documentation changes..

chidozieononiwu commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -6,44 +6,138 @@
+        /// </summary>
         LineBreak = 1,
+        /// <summary>
+        /// Regular single space.
```
We can just call it `space` if we want to, that was just the name that came to me. I did update it to `nonBreakingSpace` though on the other PR. Content token with " " value could work but that's not the design.

chidozieononiwu commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -6,44 +6,138 @@
+        /// </summary>
         TabSpace = 3,
+        /// <summary>
+        /// Use this between method parameters. Depending on user setting this would result in a single space or new line.
```
This is needed to support a new request: to allow users control how parameters render (same line vs new line). We can have language defaults and also support the indentation. The separate lines will be commendable if it has tokens with ids on it.

chidozieononiwu commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -6,44 +6,138 @@
+        public StructuredTokenKind Kind { get; set; } = StructuredTokenKind.Content;
+        
+        /// <summary>
+        /// The token value which will be displayed. Spacing tokens don't need to have value.
```
They are replaced at deserialization based on the token Kind. So they don't need to be set.

maririos replied:
added

chidozieononiwu commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -6,44 +6,138 @@
-        }
+        /// <summary>
+        /// Token Id.
+        /// Previously known as DefinitionId.
```
In the current APIView we have several cases where this is not unique.

maririos replied:
> In the current APIView 

What about Tree token? should I add the word unique here?

chidozieononiwu replied:
This does not have to be unique because the same token can occur more than once on the page.

chidozieononiwu commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -6,44 +6,138 @@
-        public StructuredTokenKind Kind { get; set; } = StructuredTokenKind.Content;
+
+        /// <summary>
+        /// Opt in or opt out boolean properties
```
You could say it that way. But my thinking is a little different. Some tags are opt in e.g. `Deprecated` since by default things are not deprecated. Some are opt out `HideFromNav` since by default nodes are show in navigation. Whatever makes it clearer is fine.

chidozieononiwu commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -6,44 +6,138 @@
+        /// <description>Deprecated: Mark a node as deprecated</description>
+        /// </item>
+        /// <item>
+        /// <description>SkipDiff: Indicate that a node should not be used in computation of diff.</description>
```
Never heard of an open Enum till right now. It does not seem like C# has such a concept, but this is fine with me.

chidozieononiwu replied:
Perhaps I can define the tags as constants.

chidozieononiwu commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -158,43 +252,113 @@ public static StructuredToken CreateStringLiteralToken(string value)
     public class APITreeNode
     {
-        public HashSet<string> Tags
+        /// <summary>
+        /// Id of the node, which should be unique at the node level. i.e. unique among its siblings.
+        /// This was previously represented by the DefinitionId for the main Token of the node.
```
You can go with either of the examples above and be fine as far as the APITreeNodes are concerned. However, you need to make sure to use the same ID used for tokens that should allow navigation to the node. And since those tokens also need to support existing APIView comments, you should just use the same style used in your existing parser i.e. Foo.Foo.bar

chidozieononiwu commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -158,43 +252,113 @@ public static StructuredToken CreateStringLiteralToken(string value)
+        public string Id { get; set; }
+
+        /// <summary>
+        /// The type of node it is. Using any of the following `assembly`, `class`, `delegate`, `enum`, `interface`,
+        /// `method` , `namespace`, `package`, `struct`, `type` will get you the corresponding default icons
+        /// for the page navigation.
+        /// </summary>
+        public string Kind { get; set; }
```
Again, fine with me although am not sure how an open Enum will work or what advantage it brings, seems like it's just more code,

chidozieononiwu replied:
I will make the defaults constants in the code.

chidozieononiwu commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -158,43 +252,113 @@ public static StructuredToken CreateStringLiteralToken(string value)
+        /// <list type="bullet">
+        /// <item>
+        /// <description>SubKind: Similar to kind, use this to make the node more specific.
+        /// e.g. `Kind = 'Type'`, and `SubKind = 'class'`. We also use this to make the navigation icon it will override kind.</description>
```
Yes, if you feel you don't need it just ignore it.

chidozieononiwu commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -158,43 +252,113 @@ public static StructuredToken CreateStringLiteralToken(string value)
         }
+
+        /// <summary>
+        /// The name of the tree node which will be used as label for the API Navigation.
```
Why? We will use the name for the navigation on the left side of the page.

maririos replied:
this was my mistake. Updated

maririos commented:
Update: I have addressed all comments that I have context about. For string vs enum vs extensible enum, we can leave those for the main PR.
@chidozieononiwu for the remaining open comments, could you please either make the changes directly on the PR or add suggestions on it for me to address them?

maririos commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -6,44 +6,138 @@
+        /// </summary>
         ParameterSeparator = 4,
+        /// <summary>
+        /// A url token should have `LinkText` property i.e `token.Properties["LinkText"]` and the url/link should be the token value.
```
@chidozieononiwu could you provide this information?

chidozieononiwu replied:
`
"value" : "the url"
"kind" : "Url",
"Properties" : [
    "LinkText" : "link text to display"
]
`

chidozieononiwu commented on src/dotnet/APIView/APIView/Model/TokenTreeModel.cs:
```diff
@@ -158,43 +252,113 @@ public static StructuredToken CreateStringLiteralToken(string value)
         }
+
+        /// <summary>
+        /// Properties of a node.
+        /// <list type="bullet">
+        /// <item>
+        /// <description>SubKind: Similar to kind, use this to make the node more specific.
+        /// e.g. `Kind = 'Type'`, and `SubKind = 'class'`. We also use this to make the navigation icon it will override kind.</description>
+        /// </item>
+        /// <item>
+        /// <description>IconName: Use this only if you are looking to add a custom icon different from language wide defaults. New additions will need to be supported APIView side.</description>
+        /// </item>
+        /// </list>
```
I will make the defaults constants in the code.
